No details given when POST fails in HttpOutput

## Summary

when HttpOutput is not able to POST the result, we don't get enough details in logs to investigate

## Type of Issue
<!-- This issue is a -->
<!-- put an `x` the boxe that apply. -->
It is a : 
 - [x] bug

## Current Behavior

This is the log output I get : 

~~~
Caused by: org.springframework.web.client.HttpClientErrorException: 400 Bad Request
12:23:44 	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:94) ~[spring-web-5.0.8.RELEASE.jar!/:5.0.8.RELEASE]
12:23:44 	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:79) ~[spring-web-5.0.8.RELEASE.jar!/:5.0.8.RELEASE]
12:23:44 	at org.springframework.web.client.ResponseErrorHandler.handleError(ResponseErrorHandler.java:63) ~[spring-web-5.0.8.RELEASE.jar!/:5.0.8.RELEASE]
12:23:44 	at org.springframework.web.client.RestTemplate.handleResponse(RestTemplate.java:766) ~[spring-web-5.0.8.RELEASE.jar!/:5.0.8.RELEASE]
12:23:44 	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:724) ~[spring-web-5.0.8.RELEASE.jar!/:5.0.8.RELEASE]
12:23:44 	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:680) ~[spring-web-5.0.8.RELEASE.jar!/:5.0.8.RELEASE]
12:23:44 	at org.springframework.web.client.RestTemplate.postForEntity(RestTemplate.java:466) ~[spring-web-5.0.8.RELEASE.jar!/:5.0.8.RELEASE]
12:23:44 	at com.societegenerale.githubcrawler.output.HttpOutput.output(HttpOutput.kt:34) ~[github-crawler-core-1.0.8.jar!/:1.0.8]
12:23:44 	at com.societegenerale.githubcrawler.GitHubCrawler$publish$1.accept(GitHubCrawler.kt:179) ~[github-crawler-core-1.0.8.jar!/:1.0.8]
~~~

However, looking at the code this is expected to be taken care of : https://github.com/societe-generale/github-crawler/blob/05bc325357446a0ccf78c85d5ab93539ca945e56/github-crawler-core/src/main/kotlin/com/societegenerale/githubcrawler/output/HttpOutput.kt#L36-L38 

But since an exception is thrown, the block check the reposnse status is never reached, and details of failure are not provided in logs 

## Expected Behavior

In addition to the error code, we need to have the error response body in the logs. 

## Your Environment

<!--- If you're reporting a bug, include as many relevant details about the environment you experienced the bug in -->
* Version used: 1.0.8

--- a/github-crawler-autoconfigure/src/main/kotlin/com/societegenerale/githubcrawler/config/GitHubCrawlerOutputConfig.kt
+++ b/github-crawler-autoconfigure/src/main/kotlin/com/societegenerale/githubcrawler/config/GitHubCrawlerOutputConfig.kt
@@ -10,6 +10,7 @@ import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
 import org.springframework.boot.context.properties.EnableConfigurationProperties
 import org.springframework.context.annotation.Bean
 import org.springframework.context.annotation.Configuration
+import org.springframework.web.client.RestTemplate
 import java.io.IOException
 
 @Configuration
@@ -33,7 +34,7 @@ open class GitHubCrawlerOutputConfig {
     @AutoConfigureOrder(value = 2)
     open fun httpOutput(@Value("\${output.http.targetUrl:TARGET_URL_IS_MANDATORY}") targetUrl: String): GitHubCrawlerOutput {
 
-        return HttpOutput(targetUrl)
+        return HttpOutput(targetUrl, RestTemplate())
     }
 
 

--- a/github-crawler-core/src/main/kotlin/com/societegenerale/githubcrawler/output/HttpOutput.kt
+++ b/github-crawler-core/src/main/kotlin/com/societegenerale/githubcrawler/output/HttpOutput.kt
@@ -2,18 +2,15 @@ package com.societegenerale.githubcrawler.output
 
 import com.societegenerale.githubcrawler.model.Repository
 import org.slf4j.LoggerFactory
+import org.springframework.http.ResponseEntity
+import org.springframework.web.client.HttpClientErrorException
 import org.springframework.web.client.RestTemplate
 
 
-class HttpOutput(private val targetUrl: String) : GitHubCrawlerOutput {
+class HttpOutput(private val targetUrl: String,
+                 private val restTemplate: RestTemplate) : GitHubCrawlerOutput {
 
     val log = LoggerFactory.getLogger(this.javaClass)
-
-    private val restTemplate: RestTemplate
-
-    init {
-        this.restTemplate = RestTemplate()
-    }
 
     override fun output(analyzedRepository: Repository) {
 
@@ -31,13 +28,22 @@ class HttpOutput(private val targetUrl: String) : GitHubCrawlerOutput {
                     analyzedRepository.ownerTeam ?: "Undefined",
                     analyzedRepository.topics)
 
-            val response = restTemplate.postForEntity(targetUrl, output, String::class.java)
+            val response : ResponseEntity<String>
 
-            if (!response.statusCode.is2xxSuccessful) {
-                log.warn("couldn't push result for repo {} - code {} - {}", analyzedRepository.name, response.statusCode, response.body)
+            try {
+                response = restTemplate.postForEntity(targetUrl, output, String::class.java)
+
+                if (!response.statusCode.is2xxSuccessful) {
+                    logHttpError(analyzedRepository.name, response.statusCodeValue,response.body)
+                }
+            } catch (e : HttpClientErrorException) {
+                logHttpError(analyzedRepository.name, e.rawStatusCode,e.message ?: "no additional message provided")
             }
-
         }
+    }
+
+    private fun logHttpError(repoName: String, errorCode : Int, message : String) {
+        log.warn("couldn't push result for repo {} - code {} - {}", repoName, errorCode, message)
     }
 
 }

--- a/github-crawler-core/src/test/java/com/societegenerale/githubcrawler/output/HttpOutputTest.kt
+++ b/github-crawler-core/src/test/java/com/societegenerale/githubcrawler/output/HttpOutputTest.kt
@@ -1,0 +1,78 @@
+package com.societegenerale.githubcrawler.output
+
+
+import ch.qos.logback.classic.spi.ILoggingEvent
+import ch.qos.logback.core.Appender
+import com.nhaarman.mockito_kotlin.argumentCaptor
+import com.nhaarman.mockito_kotlin.eq
+import com.nhaarman.mockito_kotlin.mock
+import com.societegenerale.githubcrawler.model.Branch
+import com.societegenerale.githubcrawler.model.Repository
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Test
+import org.mockito.Mockito.*
+import org.slf4j.LoggerFactory
+import org.springframework.http.HttpStatus
+import org.springframework.web.client.HttpClientErrorException
+import org.springframework.web.client.RestTemplate
+import java.util.*
+
+
+class HttpOutputTest{
+
+    private val logger = LoggerFactory.getLogger(HttpOutput::class.java!!)
+
+
+    @Test
+    fun shouldLogResponseBodyWhenErrorDuringPost(){
+        //Mock logging infra
+        val root = LoggerFactory.getLogger(ch.qos.logback.classic.Logger.ROOT_LOGGER_NAME) as ch.qos.logback.classic.Logger
+        val mockAppender : Appender<ILoggingEvent> = mock()
+
+        `when`(mockAppender.name).thenReturn("MOCK")
+        root.addAppender(mockAppender)
+
+
+        val mockRestTemplate : RestTemplate = mock()
+
+        val httpOutput = HttpOutput("http://localhost",mockRestTemplate)
+
+        `when`(mockRestTemplate.postForEntity(anyString(),
+                                              any(OutputIndicator::class.java),
+                                              eq(String::class.java)
+                                             )
+             ).thenThrow(HttpClientErrorException(HttpStatus.BAD_REQUEST,"Bad Request details"))
+
+
+        val dummyRepo=Repository(name = "repo1",
+                creationDate = Date(),
+                config= null,
+                defaultBranch = "master",
+                fullName = "orgName/repoName1",
+                lastUpdateDate = Date(),
+                ownerTeam = null,
+                reason = null,
+                url="http://hello",
+                branchesToParse = listOf(Branch("master"))
+        )
+
+        //ACTION
+        httpOutput.output(dummyRepo)
+
+
+        argumentCaptor<ILoggingEvent>().apply {
+            verify(mockAppender).doAppend(capture())
+
+            assertThat(allValues).isNotEmpty
+
+//            val logMessage = firstValue as String
+//         //   val command = (firstValue as GenericCommandMessage<SaveSyndicationFacilityCommand>).payload
+//            assertThat(logMessage).isEqualTo("Bad Request details")
+        }
+
+    }
+
+
+
+}
+

--- a/github-crawler-core/src/test/java/com/societegenerale/githubcrawler/output/HttpOutputTest.kt
+++ b/github-crawler-core/src/test/java/com/societegenerale/githubcrawler/output/HttpOutputTest.kt
@@ -18,61 +18,53 @@ import org.springframework.web.client.RestTemplate
 import java.util.*
 
 
-class HttpOutputTest{
+class HttpOutputTest {
 
     private val logger = LoggerFactory.getLogger(HttpOutput::class.java!!)
 
 
     @Test
-    fun shouldLogResponseBodyWhenErrorDuringPost(){
+    fun shouldLogResponseBodyWhenErrorDuringPost() {
         //Mock logging infra
         val root = LoggerFactory.getLogger(ch.qos.logback.classic.Logger.ROOT_LOGGER_NAME) as ch.qos.logback.classic.Logger
-        val mockAppender : Appender<ILoggingEvent> = mock()
+        val mockAppender: Appender<ILoggingEvent> = mock()
 
         `when`(mockAppender.name).thenReturn("MOCK")
         root.addAppender(mockAppender)
 
 
-        val mockRestTemplate : RestTemplate = mock()
+        val mockRestTemplate: RestTemplate = mock()
 
-        val httpOutput = HttpOutput("http://localhost",mockRestTemplate)
+        val httpOutput = HttpOutput("http://localhost", mockRestTemplate)
 
         `when`(mockRestTemplate.postForEntity(anyString(),
                                               any(OutputIndicator::class.java),
                                               eq(String::class.java)
                                              )
-             ).thenThrow(HttpClientErrorException(HttpStatus.BAD_REQUEST,"Bad Request details"))
+              ).thenThrow(HttpClientErrorException(HttpStatus.BAD_REQUEST, "Bad Request details"))
 
-
-        val dummyRepo=Repository(name = "repo1",
+        val dummyRepo = Repository(name = "repo1",
                 creationDate = Date(),
-                config= null,
+                config = null,
                 defaultBranch = "master",
                 fullName = "orgName/repoName1",
                 lastUpdateDate = Date(),
                 ownerTeam = null,
                 reason = null,
-                url="http://hello",
+                url = "http://hello",
                 branchesToParse = listOf(Branch("master"))
         )
 
         //ACTION
         httpOutput.output(dummyRepo)
 
-
         argumentCaptor<ILoggingEvent>().apply {
             verify(mockAppender).doAppend(capture())
 
-            assertThat(allValues).isNotEmpty
+            val matchingLogEvent = allValues.firstOrNull{ logEvent -> logEvent.formattedMessage.contains("Bad Request details") }
 
-//            val logMessage = firstValue as String
-//         //   val command = (firstValue as GenericCommandMessage<SaveSyndicationFacilityCommand>).payload
-//            assertThat(logMessage).isEqualTo("Bad Request details")
+            assertThat(matchingLogEvent ).`as`("can't find a log statement with expected message").isNotNull
         }
-
     }
-
-
-
 }
 

--- a/pom.xml
+++ b/pom.xml
@@ -149,6 +149,23 @@
             <version>${kotlin.version}</version>
             <scope>test</scope>
         </dependency>
+
+        <dependency>
+            <groupId>com.nhaarman</groupId>
+            <artifactId>mockito-kotlin</artifactId>
+            <scope>test</scope>
+            <version>1.6.0</version>
+            <exclusions>
+                <exclusion>
+                    <groupId>org.jetbrains.kotlin</groupId>
+                    <artifactId>kotlin-stdlib</artifactId>
+                </exclusion>
+                <exclusion>
+                    <groupId>org.jetbrains.kotlin</groupId>
+                    <artifactId>kotlin-reflect</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
     </dependencies>
 
     <build>

